### ISSUE TITLE
Do not call sys.exc_clear() on Python 3 in twisted/trial

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8469]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)| @rodrigc reported|
|-|-|
|Trac ID|trac#8469|
|Type|enhancement|
|Created|2016-06-13 22:00:15Z|
|Branch|https://github.com/twisted/twisted/tree/8469-rodrigc-exc_clear|

sys.exc_clear() is gone on Python 3: https://docs.python.org/3/whatsnew/3.0.html#index-22

We should not try to call it on Python 3.

<details><summary>Searchable metadata</summary>

```
trac-id__8469 8469
type__enhancement enhancement
reporter__rodrigc rodrigc
priority__normal normal
milestone__Python_3_x Python-3.x
branch__8469_rodrigc_exc_clear 8469-rodrigc-exc_clear
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__None None
time__1465855215565071 1465855215565071
changetime__1467195411094770 1467195411094770
version__None None
owner__rodrigc rodrigc
cc__jml
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_2"></a>|@rodrigc commented|
|-|-|

See:
https://github.com/twisted/twisted/pull/188

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_4"></a>|@Lukasa set owner to @rodrigc|
|-|-|

Here's some review that applies not only to this but to [#8465](https://github.com/twisted/twisted/issues/8465), [#8466](https://github.com/twisted/twisted/issues/8466), [#8467](https://github.com/twisted/twisted/issues/8467), and [#8468](https://github.com/twisted/twisted/issues/8468): what the hell is `exc_clear()` even doing there? It seems to serve no purpose whatsoever. Glyph has suggested it has something to do with a Python 2.2 implementation detail, so I'm inclined to say this is crazy-talk.

I'm going to let this build run, but I'd like to see one of two things:

- Either more aggressive patch that simply changes the except block to `pass`.
- Or the most aggressive patch, which removes all of this nonsense and makes these all `console_scripts`, such that `pip install -e '.[dev]'` (the command we tell people to run for development anyway) makes those scripts available directly.

I'll happily accept either kind of patch, but we should definitely do one of them.

I'll apply this feedback to the other tickets as well. Feel free to ping me directly when you're ready for more review.

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_5"></a>|@rodrigc commented|
|-|-|

I updated the branch https://github.com/twisted/twisted/compare/trunk...8469-rodrigc-exc_clear

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)|@rodrigc set owner to @Lukasa|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_7"></a>|@rodrigc commented|
|-|-|

Lukasa, sys.exc_clear() is still needed on Python 2 in this particular part of trial.
If I take it out, then I get this test failure:

```
[FAIL]
Traceback (most recent call last):
  File "../../../twisted/trial/_dist/test/test_workertrial.py", line 136, in test_readInterrupted
    self.assertEqual([(None, None, None)], excInfos)
  File "/Users/crodrigues/twisted9/build/py27-tests-posix/lib/python2.7/site-packages/twisted/trial/_synctest.py", line 425, in assertEqual
    super(_Assertions, self).assertEqual(first, second, msg)
  File "/usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/case.py", line 515, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/case.py", line 744, in assertListEqual
    self.assertSequenceEqual(list1, list2, msg, seq_type=list)
  File "/usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/unittest/case.py", line 726, in assertSequenceEqual
    self.fail(msg)
twisted.trial.unittest.FailTest: Lists differ: [(None, None, None)] != [(<type 'exceptions.IOError'>,...

First differing element 0:
(None, None, None)
(<type 'exceptions.IOError'>, IOError(4,), <traceback object at 0x10e40bab8>)

- [(None, None, None)]
+ [(<type 'exceptions.IOError'>, IOError(4,), <traceback object at 0x10e40bab8>)]

test_workertrial.MainTests.test_readInterrupted
```

so the patch must be as I have submitted it.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_8"></a>|@Lukasa commented|
|-|-|

I think we should fix individual tests, rather than hiding this in trial. This test can just have exc_clear() added to it as we've done before.

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_9"></a>|@rodrigc commented|
|-|-|

Would this work for you:

```
diff --git a/twisted/trial/_dist/test/test_workertrial.py b/twisted/trial/_dist/test/test_workertrial.py
index b0059eb..c282e3e 100644
--- a/twisted/trial/_dist/test/test_workertrial.py
+++ b/twisted/trial/_dist/test/test_workertrial.py
@@ -121,6 +121,8 @@ class MainTests(TestCase):
             count = 0
 
             def read(oself, size):
+                if sys.version_info < (3, 0):
+                    sys.exc_clear()
                 oself.count += 1
                 if oself.count == 1:
                     raise IOError(errno.EINTR)
diff --git a/twisted/trial/_dist/workertrial.py b/twisted/trial/_dist/workertrial.py
index 0cdc43b..3671d5c 100644
--- a/twisted/trial/_dist/workertrial.py
+++ b/twisted/trial/_dist/workertrial.py
@@ -85,7 +85,6 @@ def main(_fdopen=os.fdopen):
             r = protocolIn.read(1)
         except IOError as e:
             if e.args[0] == errno.EINTR:
-                sys.exc_clear()
                 continue
             else:
                 raise
```

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_10"></a>|@Lukasa commented|
|-|-|

Sorry rodrigc, but having seen the diff I understand better what you were getting at. Yeah, it might be better to replace the exc_clear in its original spot in this case.

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_11"></a>|@Lukasa set owner to @rodrigc|
|-|-|

Ok, LGTM. Merge when ready.

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|GitHub set status to `closed`|
|-|-|

In changeset b3ece589e2809d41ad589b1ec7dee0587bd9af18
```
#!CommitTicketReference repository="" revision="b3ece589e2809d41ad589b1ec7dee0587bd9af18"
Merge 8469-rodrigc-exc_clear-2: Do not call sys.exc_clear() on Python 3 in twisted/trial

Author: rodrigc
Reviewer: Lukasa
Fixes: #8469
```
